File: 2018/09/rs/src/main.rs
use regex::Regex;
use std::collections::VecDeque;

fn solve(puzzle_input: &(usize, usize)) -> (usize, usize) {
    let (elves_count, last_marble) = *puzzle_input;
    let mut part1_score = 0;
    let mut scores = vec![0; elves_count];
    let mut circle = VecDeque::with_capacity(last_marble * 100);
    circle.push_back(0);
    for next_marble in 1..=last_marble * 100 {
        if next_marble == last_marble {
            part1_score = *scores.iter().max().unwrap();
        }
        if next_marble % 23 == 0 {
            for _ in 0..7 {
                let previous = circle.pop_back().unwrap();
                circle.push_front(previous);
            }
            scores[next_marble % elves_count] += next_marble + circle.pop_front().unwrap();
        } else {
            for _ in 0..2 {
                let next = circle.pop_front().unwrap();
                circle.push_back(next);
            }
            circle.push_front(next_marble);
        }
    }
    (part1_score, *scores.iter().max().unwrap())
}

fn get_input(file_path: &String) -> (usize, usize) {
    let re = Regex::new(r"^(?P<players>\d+) players; last marble is worth (?P<last>\d+)").unwrap();
    re.captures(
        std::fs::read_to_string(file_path)
            .expect("Error reading input file!")
            .as_str(),
    )
    .map(|cap| {
        (
            cap.name("players").unwrap().as_str().parse().unwrap(),
            cap.name("last").unwrap().as_str().parse().unwrap(),
        )
    })
    .unwrap()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    let end = now.elapsed().as_secs_f32();
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", end);
}

File: 2018/09/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>

typedef struct {
    int elvesCount, lastMarble;
} Input;
typedef struct
{
    long part1;
    long part2;
} Results;

typedef struct {
    int previous, next;
} Marble;

typedef struct {
    int current;
    Marble *marbles;
} Circle;

long max(long *scores, int size)
{
    long result = 0;
    while (size--)
    {
        result = *scores > result ? *scores : result;
        scores++;
    }
    return result;
}

void printCircle(Circle circle)
{
    int marble = 0;
    do
    {
        if (marble == circle.current)
            printf("(");
        printf("%d", marble);
        if (marble == circle.current)
            printf(")");
        printf(" ");
        marble = circle.marbles[marble].next;
    } while (marble);
}

Results solve(Input input)
{
    long *scores = calloc(input.elvesCount, sizeof(long));
    Circle circle = {0, malloc(input.lastMarble * 100 * sizeof(Marble))};
    circle.marbles[0] = (Marble){0, 0};
    long part1Score = 0, rotate;
    for (int nextMarble = 1; nextMarble <= input.lastMarble * 100; nextMarble++)
    {
        if (nextMarble == input.lastMarble)
            part1Score = max(scores, input.elvesCount);
        if (nextMarble % 23)
        {
            circle.current = circle.marbles[circle.current].next;
            circle.marbles[nextMarble] = (Marble){circle.current, circle.marbles[circle.current].next};
            circle.marbles[circle.marbles[circle.current].next].previous = nextMarble;
            circle.marbles[circle.current].next = nextMarble;
            circle.current = nextMarble;
        }
        else
        {
            for (rotate = 0; rotate < 7; rotate++)
                circle.current = circle.marbles[circle.current].previous;
            scores[nextMarble % input.elvesCount] += nextMarble + circle.current;
            circle.marbles[circle.marbles[circle.current].previous].next = circle.marbles[circle.current].next;
            circle.current = circle.marbles[circle.current].next;
        }
    }
    return (Results){part1Score, max(scores, input.elvesCount)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    int elvesCount, lastMarble;
    sscanf(content, "%d players; last marble is worth %d points", &elvesCount, &lastMarble);
    fclose(file);
    return (Input) {elvesCount, lastMarble};
}

void freeInput(Input input)
{
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %ld\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
